src/mame/drivers/swtpc8212.cpp

```diff
@@ -22,82 +60,359 @@ class swtpc8212_state : public driver_device
+
+	// TODO:
+	// PA7 - cursor bit 0
+	// PA6 - cursor bit 1
+	// PA5 - cursor bit 2
+	// PA4 - cursor bit 3
+
+	return jumper_a | jumper_b;
+}
+
+READ8_MEMBER(swtpc8212_state::pia0_pb_r)
+{
+	return m_kbd_data;
+}
+
+void swtpc8212_state::kbd_put(uint8_t data)
+{
+	m_kbd_data = data;
+	// Triggers on the falling edge.
+	m_pia0->cb1_w(ASSERT_LINE);
+	m_pia0->cb1_w(CLEAR_LINE);
+	m_pia0->cb1_w(ASSERT_LINE);
+}
+
+WRITE_LINE_MEMBER(swtpc8212_state::pia0_ca2_w)
+{
+	if (state == 0)
+	{
+		m_beeper->set_state(1);
+		m_bell_timer->reset(attotime::from_msec(250));
+	}
+}
+
+WRITE8_MEMBER(swtpc8212_state::pia1_pa_w)
+{
+	// External parallel printer data output.
+	m_printer_data = data;
+}
+
+READ_LINE_MEMBER(swtpc8212_state::pia1_ca1_r)
+{
+	// External parallel printer busy input.
+	return 0;
+}
+
+WRITE_LINE_MEMBER(swtpc8212_state::pia1_ca2_w)
+{
+	// External parallel printer data ready.
+	// Trigger on the falling edge.
+	if (m_printer_data_ready == 1 && state == 0)
+	{
+		m_printer->output(m_printer_data);
+		// Toggle the printer busy line as the software waits for a
+		// falling edge.
+		m_pia1->ca1_w(CLEAR_LINE);
+		m_pia1->ca1_w(ASSERT_LINE);
+		m_pia1->ca1_w(CLEAR_LINE);
+	}
+	m_printer_data_ready = state;
+}
+
+READ8_MEMBER(swtpc8212_state::pia1_pb_r)
+{
+	uint8_t baud_rate = m_baud_rate->read(); // PB0 to PB4
+	uint8_t mode = m_mode_switch->read();    // PB5
+	uint8_t no_parity = m_no_parity->read(); // PB6
+	uint8_t parity_select = m_parity_select->read(); // PB7
+
+	return parity_select | no_parity | mode | baud_rate;
+}
+
+
 MC6845_UPDATE_ROW(swtpc8212_state::update_row)
 {
+	int x = 0;
+	uint8_t *chargen = BIT(m_latch_data, 6) == 0 ? m_chargen1 : m_chargen2;
+
+	for (int column = 0; column < x_count; column++)
+	{
+		uint8_t code = m_video_ram[(ma + column) & 0x07ff];
+		int dcursor = (column == cursor_x);
+
+		offs_t address = ra < 16 ? (code & 0x7f) | (ra & 0x0f) << 7 : 0;
+		uint8_t data = chargen[address];
+		uint8_t intensity = BIT(code, 7);
+
+		for (int bit = 0; bit < 8; bit++)
+		{
+			int dout = BIT(data, 7);
+			uint32_t font_color = 0;
+			if ((dcursor ^ dout) && de)
+			{
+				if (intensity)
+					font_color = rgb_t(0x10, 0xff, 0x10);
+				else
+					font_color = rgb_t(0x00, 0xd0, 0x00);
+			}
+			bitmap.pix32(y, x++) = font_color;
+			data <<= 1;
+		}
+
+		// Gap between characters
+		if (BIT(m_latch_data, 7) == 0)
+			x++;
+	}
+}
+
+void swtpc8212_state::device_timer(emu_timer &timer, device_timer_id id, int param, void *ptr)
+{
+	switch (id)
+	{
+	case BELL_TIMER_ID:
+		m_beeper->set_state(0);
+		break;
+	}
 }
 
+void swtpc8212_state::machine_start()
+{
+	m_bell_timer = timer_alloc(BELL_TIMER_ID);
+
+	save_item(NAME(m_latch_data));
+	save_item(NAME(m_kbd_data));
+	save_item(NAME(m_printer_data));
+	save_item(NAME(m_printer_data_ready));
+}
+
+void swtpc8212_state::machine_reset()
+{
+	m_kbd_data = 0;
+	m_pia0->cb1_w(ASSERT_LINE);
+
+	m_latch_data = 0x1f;
+
+	m_uart->cts_w(CLEAR_LINE);
```

Comment:
This porbably shouldn't be here - CTS comes from the device attached to the port, and if that device is keeping CTS deasserted when the terminal is rese, the input won't magically change.

Reply:
Most classic UARTs and USARTs automatically and unconditionally disable the transmitter when CTS is in a marking state (1). Many system designers disliked this behavior and tied the CTS pin to GND, with the actual RS232 CTS being read from some other port (e.g. VT100).